Adding Iterator and remane functions (#148)
ARM-software/psa-api: doc/storage/api/api.rst

The |API| uses the status code definitions that are shared with the other PSA Certified APIs.

> The following elements are defined in :file:`psa/error.h` from :cite-title:`PSA-STAT` (previously defined in :cite:`PSA-FF-M`):

Reviewer: This is another artifact of the initial file content not being fully synchronised with upstream.

Reviewer: This has been broken again. Related to the previous commit about build errors which also undoes changes made in main in January,

    .. field:: psa_storage_create_flags_t flags
        The flags set when the ``uid`` was create

.. struct:: psa_storage_iterator_t

    .. summary::

>         An implantation defined opaque structure containing the context for an iterator. 

Reviewer: ```suggestion
        An implementation-defined opaque structure containing the context for an iterator. 
```

        The structure MUST  contain all all the state required by the iterator. 
        No further state is retained by the implementation. 
        
        The structure is initilaised by the `ps_iterator_start` function. 
        It is modified by the `ps_iterator_next` function. 
        

>         It is the duty of the calling application to free memory when the iterator is no longer required. 

Reviewer: As there is no hidden allocation, using the term 'free' might be confusing. Might it be better to state that the application can discard or reuse the iterator object once it has finished using it. This can be before, or after, the iterator has reached the end of the iteration.

    It will be incremented in small updates that are unlikely to include breaking changes.

.. macro:: PSA_ITS_SUPPORT_ITERATION
    (1u << 2)

    Flag indicating that `psa_its_iterator_start()` and  `psa_its_iterator_next()` are supported.

> .. struct:: psa_its_storage_iterator_t

Reviewer: I've changed my mind: as this is a fully implementation-defined type, we should probably use the same pattern as the Crypto API for such types, and just use a typedef in this specification. For example, see [the definition of `psa_key_attributes_t`](https://arm-software.github.io/psa-api/crypto/1.2/api/keys/attributes.html#c.psa_key_attributes_t). This gives the implementation full flexibility in the definition of the type.

To do this, replace:

```
.. struct:: psa_its_storage_iterator_t
    :type:
```

with

```
.. typedef:: /* implementation-defined type */ psa_its_storage_iterator_t
```

Likewise for the PS iterator type.

In the Crypto API, we then refer to these as 'objects' rather than 'structures' - i.e. use 'iterator object' in this spec when describing a variable of this type.

    .. retval:: PSA_ERROR_STORAGE_FAILURE
        The operation failed because the physical storage has failed (Fatal error).

>    Introduced in Version 1.1. 

Reviewer: `psa_its_remove()` has been in the API since 1.0. Lines 335-337 are not required here.

.. function:: psa_its_rename

   .. summary::
      Atomically renames the storage location with the specified ``uid`` to a ``uid_new``.

   .. param:: psa_storage_uid_t uid
        The current identifier for the data.

   .. param:: psa_storage_uid_t uid_new
        The new identifier for the data.

   .. param:: psa_storage_rename_flags_t rename_flags
        The flags must be either ``PSA_STORAGE_FLAG_NONE`` or ``PSA_STORAGE_FLAG_REPLACE``

   .. return:: psa_status_t
        A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
        The operation completed successfully.

   .. retval:: PSA_ERROR_ALREADY_EXISTS
        Storage with the specified ``uid_new`` already exists and ``rename_flags`` is `PSA_STORAGE_FLAG_NONE`

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
        Storage with the specified ``uid`` does not exist.

   .. retval:: PSA_ERROR_GENERIC_ERROR
        The operation failed because of an unspecified internal failure.

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
        The operation failed because either:

        * ``uid`` is ``0``.
        * ``uid_new`` is ``0``

>         * the ``psa_storage_rename_flags_t`` has a value set other than `PSA_STORAGE_FLAG_REPLACE`

Reviewer: This should be about the `rename_flags` parameter, not the type.

```suggestion
        * ``rename_flags`` has a value set other than `PSA_STORAGE_FLAG_REPLACE`.
```

   .. param::  psa_its_storage_iterator_t* context
       A pointer to a context for this iterator as returned by `psa_its_iterator_start()` or updated by a previous call to `psa_its_iterator_next()`.  The content of the iterator will change on success and is undefined on error.

   .. param:: psa_storage_uid_t *result
        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the next matching ``uid``. On error, the contents are undefined.

   .. return:: psa_status_t
       A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
       The operation completed successfully.

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
       The iterator has returned all the uids that match this iteration.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
       The operation failed because either:

       * The provided context is not valid.

       * The caller cannot access the memory at ``result``

   Introduced in Version 1.1. 
   

>    This function takes an iterator that was returned by the `psa_its_iterator_start` function and searches the storage for the next ``uid`` that matches the filters defined in that function call. If a ``uid`` matching the filter exists, the function updates the ``result`` and ``context`` parameters and retursn PSA_SUCCESS. If not it reruns an error.        

Reviewer: Spelling: 'reruns' -> 'returns'

      This function must be fully defined if iteration is supported.
      If iteration is not supported, then this function SHALL be defined in the header - potentially as an inline function - but SHALL return ``PSA_ERROR_NOT_SUPPORTED``

   .. param::  psa_its_storage_iterator_t* context
       A pointer to a context for this iterator as returned by `psa_its_iterator_start()` or updated by a previous call to `psa_its_iterator_next()`.  The content of the iterator will change on success and is undefined on error.

   .. param:: psa_storage_uid_t *result
        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the next matching ``uid``. On error, the contents are undefined.

   .. return:: psa_status_t
       A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
       The operation completed successfully.

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
       The iterator has returned all the uids that match this iteration.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).

>    .. retval:: PSA_ERROR_DATA_CORRUPT

Reviewer: This contradicts the description text relating to the stability of the iterator when the underlying data storage is modified by the application between calls?

       The operation failed because the contents of the iteration have changed. That is a ``uid`` matching the filter has either been created or deleted.

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
       The operation failed because either:

       * The provided context is not valid.

       * The caller cannot access the memory at ``result``

>    The iterator returns those values where the ``filter_length`` bits of the ``uid`` matches the left most bits in ``filter``.

Reviewer: This and the following descriptive text would be better placed in the `psa_its_iterator_start()` function. This function can refer to the start function for description of the behavior of the iterator if the storage is modified between iterator calls.

    * Even if all parameters are correct, the function can fail in the case of a storage failure.

.. function:: psa_its_iterator_start

   .. summary::
       Initializes an iterator that can be used to return a list of ``uid`` values in the Internal Trusted Storage.
       
       This function must be fully defined if `PSA_STORAGE_SUPPORT_ITERATION` is true.
       
       If `PSA_STORAGE_SUPPORT_ITERATION` is false, then this function SHALL always return ``PSA_ERROR_NOT_SUPPORTED``

   .. param::  psa_its_storage_iterator_t* context

>        A pointer to a context for this iterator. The pointer may be NULL. This is set to a new value on success and is undefined on error. The content of the iterator is implementation defined.

Reviewer: The optional `NULL` value looks odd, as it would seem that the entire purpose of starting an iteration would be to iterate.

There is no mention of how the function behaves if no iterator object is provided, or why a caller would ever want to do this. If there is a use case (I can think of a couple) where this is a better approach than provided an iterator object, it would benefit from an explanation and/or code example in the description.

Reviewer: The same applies to `psa_ps_iterator_start()`.

        * The ``uid`` is ``0``.

        * The operation failed because caller cannot access some or all of the memory in the range [``p_data``, ``p_data + data_length - 1``].
        

>         * the uid exists and ``data_length`` is greater then ```capacity`` 

Author: I think the intention was always that if a file was created with capacidy = x then any set with size < x will succeed. 

I thought the initial intention was that if size > capacity - for either set or set_extended - then it fials. 

If the intention was that it might succeed, but cannot be guaranteed, that is also fine.

>         * the uid exists and ``data_length`` is greater then ```capacity`` 

Reviewer 2: > I think the intention was always that if a file was created with capacity = x then any set with size < x will succeed.

I don't know what the intent was, but I find this requirement surprising. In a log-structured filesystem, overwriting a file with less content takes some disk space. Even deleting a file can require some disk space in some filesystems.

    It will be incremented in small updates that are unlikely to include breaking changes.

.. macro:: PSA_PS_SUPPORT_ITERATION
    (1u << 2)

    Flag indicating that `psa_ps_iterator_start` and `psa_ps_iterator_next` are supported.

> .. struct:: psa_ps_storage_iterator_t

Reviewer: Make this a typedef, as per the ITS iterator:

```suggestion
.. struct:: psa_ps_storage_iterator_t
    :type:
```

    .. summary::
        An implementation-defined opaque structure containing the context for an iterator.
        The structure MUST contain all all the state required by the iterator.
        That is, further state MUST NOT be retained by the implementation.

        The structure is initilaised by the ``ps_iterator_start()`` function.
        It is modified by the ``ps_iterator_next()`` function.

>         the caller can discard or reuse the iterator object once it has finished using it. This can be before, or after, the iterator has reached the end of the iteration.

Reviewer: ```suggestion
        The caller can discard or reuse the iterator object once it has finished using it. This can be before, or after, the iterator has reached the end of the iteration.
```

>         The header file is only required to define this structure if PSA_STORAGE_SUPPORT_ITERATION is true.

Reviewer: Discussion on optionality as per ITS iterator.

        * ``uid_new`` exists and ``PSA_STORAGE_FLAG_OVERWRITE`` is not specified

        * ``uid`` was created with ``PSA_STORAGE_FLAG_NO_CONFIDENTIALITY`` and ``uid_new`` exists and does not have that flag set
        * ``uid`` was created with ``PSA_STORAGE_FLAG_NO_REPLAY_PROTECTION`` and ``uid_new`` exists and does not have that flag set

   .. retval:: PSA_ERROR_NOT_SUPPORTED
        The implementation does not support the operation.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
        The operation failed because the physical storage has failed (Fatal error).     

   The function renames ``uid`` to ``uid_new`` retaining the storage flags  that ``uid`` was created with. 

   If the caller specifies ``PSA_STORAGE_FLAG_OVERWRITE`` the operation atomically replaces the exisitng contents of ``uid_new`` with those of ``uid``.

   However, ``uid`` must not have been created with more permissive Storage flags than ``uid_new``
        
   Except in the case of ``PSA_ERROR_STORAGE_FAILURE``, in which case no guarantees can be made, the operation shall either succeed or leave storage unchanged. 

> .. function:: psa_ps_iterator_init

Reviewer: The iteration API needs a 'usage' example to show how applications are expected to use the interfaces, and what kinds of operation can be done while iterating.

It must also be very clearly documented that there is no 'close'/'resource release' function for iterators
I think this aspect of the API design should be much more clearly stated, as it has a critical effect on the way applications use them, and on the possible implementation choices. In particular, a RoT service implementation cannot afford to have service-side state relating to an iterator, given that:

* An application is permitted to have more than one active iterator
* It is not assumed or required that the application always operates an iterator (calls 'next' on it) until it reaches the end of the iteration

I think this interface design is acceptable, and implementations can maintain adequate iterator state in the iterator context (in the application) to meet the requirements of the use cases for the API.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).
   .. retval:: PSA_ERROR_INVALID_ARGUMENT
       The operation failed because the caller cannot access some or all of the memory in the range [``context``, ``context + PSA_STORAGE_ITERATOR_CTX__SIZE - 1``].    

   Any bit positions where the mask is set to one must match the value in the filter.
   
   That is, the iterator returns those values where ``uid & mask = filter``

   The iterator will only returns those ``uid`` that were created by the caller. It MUST not return any ``uid`` created by a different user. 

   An iterator is not required to return uids in any specific order, but MUST return them in a consistent order each time it is called. For example, if an implementation returns entries in numerical order, it should not arbitrarily change to returning them in creation order. However, the caller should not make assumptions as to the order in which entries are returned, except that each uid will be returned only once in each iteration.

   Changes to storage by other users MUST NOT affect any open iterations. 

>    A caller may initialize multiple iteration contexts at the same time. Each iteration shall be independent. Calling ``psa_ps_iterator_next()`` on one iterator MUST not effect any other open iteration. 

Reviewer: This is a critical part of the specification - combined with the absence of a 'resource release' function, it indicates [implicitly] that there cannot be any RoT service-side state associated with an iterator, without causing an unmanageable resource exhaustion problem.

        The new identifier for the data.

   .. param:: psa_storage_rename_flags_t rename_flags
        The flags must be either ``PSA_STORAGE_FLAG_NONE`` or ``PSA_STORAGE_FLAG_REPLACE``

   .. return:: psa_status_t
        A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
        The operation completed successfully.

   .. retval:: PSA_ERROR_ALREADY_EXISTS
        Storage with the specified ``uid_new`` already exists and ``rename_flags`` is ``PSA_STORAGE_FLAG_NONE``

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
        Storage with the specified ``uid`` does not exist.        

   .. retval:: PSA_ERROR_GENERIC_ERROR
        The operation failed because of an unspecified internal failure.

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
        The operation failed because either:

        * ``uid`` is ``0``.
        * ``uid_new`` is ``0``
        * the ``psa_storage_rename_flags_t`` has a value set other than ``PSA_STORAGE_FLAG_REPLACE``

   .. retval:: PSA_ERROR_NOT_PERMITTED
        The operation failed because either:                
        
        * ``uid_new`` exists and was created with ``PSA_STORAGE_FLAG_WRITE_ONCE``.        

>         * ``uid_new`` exists and ``PSA_STORAGE_FLAG_REPLACE`` is not specified

Reviewer: This condition is reported with `PSA_ERROR_ALREADY_EXISTS`

   .. retval:: PSA_ERROR_NOT_SUPPORTED
        The implementation does not support the operation.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
        The operation failed because the physical storage has failed (Fatal error).     

   The function renames ``uid`` to ``uid_new`` retaining the storage flags  that ``uid`` was created with. 

   If the caller specifies ``PSA_STORAGE_FLAG_REPLACE`` the operation atomically replaces the exisitng contents of ``uid_new`` with those of ``uid``.

   Except in the case of ``PSA_ERROR_STORAGE_FAILURE``, in which case no guarantees can be made, the operation shall either succeed or leave storage unchanged. 

.. function:: psa_ps_iterator_start

   .. summary:: 
       Initialises an iterator that can be used to return a list of uids in the Protected Storage .

   .. param::  psa_storage_iterator_t* context

>        A pointer to a context for this iterator. This is set to a new value on success and is undefined on error. The content of the iterator is implementation defined.  The pointer MUST be valid and MUST point to a block of memory of the size given by the ``PSA_STORAGE_ITERATOR_CTX__SIZE`` macro to which the caller has write access. 

Reviewer: The final sentence is now redundant

   .. param:: psa_storage_uid_t mask
       A value used to filter the results included in this iteration. 
   .. param:: psa_storage_uid_t filter
       A value used to filter the results included in this iteration. 
       

>    .. param:: *psa_storage_uid_t result

Reviewer: ```suggestion
   .. param:: psa_storage_uid_t *result
```

        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the first matching ``uid``. On error, the contents are undefined. 
        
   .. return:: psa_status_t
       A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
       The operation completed successfully.
   .. retval:: PSA_ERROR_DOES_NOT_EXIST
       No ``uid`` matches this iteration.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).

>    .. retval:: PSA_ERROR_INVALID_ARGUMENT

Reviewer: No longer necessary

   
   That is, the iterator returns those values where ``uid & mask = filter``

   The iterator will only returns those ``uid`` that were created by the caller. It MUST not return any ``uid`` created by a different user. 

   An iterator is not required to return uids in any specific order, but MUST return them in a consistent order each time it is called. For example, if an implementation returns entries in numerical order, it should not arbitrarily change to returning them in creation order. However, the caller should not make assumptions as to the order in which entries are returned, except that each uid will be returned only once in each iteration.

   Changes to storage by other users MUST NOT affect any open iterations. 
   
   ### todo: We need to state what effects are required/denied/permitted (i.e. imp-def) when the caller modifies the storage while iterating. ###

   A caller may initialize multiple iteration contexts at the same time. Each iteration shall be independent. Calling ``psa_ps_iterator_next()`` on one iterator MUST not effect any other open iteration. 

.. function:: psa_ps_iterator_next

>    .. summary::

Reviewer: Best to have a blank line between the `.. function::` directive and the content

>    Returns the next ``uid`` in this iteration. 

Reviewer: Requires indenting to make it part of the `.. summary::` directive

   .. param::  psa_storage_iterator_t* context
       A pointer to a context for this iterator as returned by ``psa_ps_iterator_start()`` or updated by a previous call to ``psa_ps_iterator_next``.  The content of the iterator will change on success and is undefined on error.

>    .. param:: *psa_storage_uid_t result

Reviewer: ```suggestion
   .. param:: psa_storage_uid_t *result
```

       
   .. param:: psa_storage_uid_t filter
       A value used to filter the results included in this iteration. 
       
   .. param:: int_t filter_length
       A length of the filter to use, this must be a value `-63 < filter_lemngth < 63`. 

   .. param:: psa_storage_uid_t *result
        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the first matching ``uid``. On error, the contents are undefined. 
        
   .. return:: psa_status_t
       A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
       The operation completed successfully.
       
   .. retval:: PSA_ERROR_DOES_NOT_EXIST
       No ``uid`` matches this iteration.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).

>    The iterator returns those values where the ``filter_length`` bits of the `uid` matches the bits in ``filter``. If the value of ``filter_length`` is positive, it matches left most bits, if it i nagative, it matches the right most bits. 

Reviewer: In practice, an implementation is only going to be designed to optimise one of the 'ends' of the uid value. E.g. if the data index is in order of uid value, then optimised filtering is only possible when matching on the most significant bits.

Might it be better to only permit filtering at one end of the uid, so that application usage of the uid values and filtering will always match any optimisation that an implementation can provide?

   The iterator will only returns those ``uid`` that were created by the caller. It MUST not return any ``uid`` created by a different user. 

   An iterator is not required to return uids in any specific order, but MUST return them in a consistent order each time it is called. For example, if an implementation returns entries in numerical order, it should not arbitrarily change to returning them in creation order. However, the caller should not make assumptions as to the order in which entries are returned, except that each uid will be returned only once in each iteration.

   Changes to storage by other users MUST NOT affect any open iterations. 

   A caller may initialize multiple iteration contexts at the same time. Each iteration shall be independent. Calling ``psa_ps_iterator_next()`` on one iterator MUST not effect any other open iteration. 
   
   A caller may read the contents of any `uid` with ``psa_ps_get()`` or write with ``psa_ps_set()`` or ``psa_ps_set_extended()`` without invalidating the iteration context. 

>    A caller may create a `uid` with ``psa_ps_set()`` or ``psa_ps_create()`` without invalidating the iteration context, provided the `uid` does *NOT* match the filter. However, if the `uid` matches the filter then any later call to `psa_ps_iterator_next()` fails with `PSA_ERROR_DATA_CORRUPT`.

Reviewer: These constraints may be challenging to implement. Does this requirement place an arbitrary cost on the implementation?

For example, if an application starts an iteration, and then creates 20 new data items, how does the implementation determine if it must report `PSA_ERROR_DATA_CORRUPT`? It does not have information about the iterator (all state in iterator object in caller), and storing all of the UIDs modified since last iteration, for checking on next ietration, leads to arbitrary volatile storage requirements.

>    A caller may delete a `uid` with ``psa_ps_remove(uid)`` without invalidating the iteration context, provided the `uid` does *NOT* match the filter. However, if the `uid` matches the filter then any later call to `psa_ps_iterator_next()` fails with `PSA_ERROR_DATA_CORRUPT`.

Reviewer: With this definition, an application that wants to remove all items that match the iterator, cannot do:

```
if (psa_ps_iterator_start(&iter, filter, bits, &uid)==0) do {
    psa_ps_remove(uid);
} while (psa_pw_iterator_next(&iter, &uid)==0);
```

But instead must use a less intuitive:

```
while (psa_ps_iterator_start(&iter, filter, bits, &uid)==0) {
    psa_ps_remove(uid);
}
```

More troubling are use cases where the removal is dependent on some other test, resulting in non-linear operation as the iterator has to be restarted for each item deleted.

This also complicates use cases where the matching set of items must be modified in a particular way. This can only be done by making two passes - in the first, all UIDs matching the filter are stored in application memory, and then each one can be modified.

   The iterator will only returns those ``uid`` that were created by the caller. It MUST not return any ``uid`` created by a different user. 

   An iterator is not required to return uids in any specific order, but MUST return them in a consistent order each time it is called. For example, if an implementation returns entries in numerical order, it should not arbitrarily change to returning them in creation order. However, the caller should not make assumptions as to the order in which entries are returned, except that each uid will be returned only once in each iteration.

   Changes to storage by other users MUST NOT affect any open iterations. 

   A caller may initialize multiple iteration contexts at the same time. Each iteration shall be independent. Calling ``psa_ps_iterator_next()`` on one iterator MUST not effect any other open iteration. 
   
   An iterator MUST return all data objects whose `uid` matches the filter that are extant when the filter was created, unless these are deleted or renamed before the iteration would return them, or the caller stops before all matching objects have been returned. 
   

>    A caller may delete a `uid` with ``psa_ps_remove(uid)`` without invalidating the iteration context. the iterator MUST never return a `uid` that has been deleted. However, if the caller is multi-threaded it ia possible another thread may delete a `uid`. 

Reviewer: That last sentence probably isn't required - except perhaps as a note. It does not constrain the implementation, but is just a reminder to multi-threaded app developers about the risk of race conditions.

>    A caller may delete a `uid` with ``psa_ps_remove(uid)`` without invalidating the iteration context. the iterator MUST never return a `uid` that has been deleted. However, if the caller is multi-threaded it ia possible another thread may delete a `uid`. 

Reviewer: * For inline code that includes API identifiers, using the `:code:` role enables the identifiers to be hyperlinked.
* For inline code that is _not_ an API element, use double back-quotes.
* For inline code that is _only_ an API element, use single back-quotes.

If in doubt, use the `:code:` role which will hyperlink if possible, otherwise render as code.

```suggestion
   A caller may delete a ``uid`` with :code:`psa_ps_remove(uid)` without invalidating the iteration context. the iterator MUST never return a ``uid`` that has been deleted. However, if the caller is multi-threaded it ia possible another thread may delete a ``uid``. 
```

>    A caller may read the contents of any `uid` with ``psa_ps_get()`` or write with ``psa_ps_set()`` or ``psa_ps_set_extended()`` without invalidating the iteration context. 

Reviewer: ```suggestion
   A caller may read the contents of any ``uid`` with `psa_ps_get()` or write with `psa_ps_set()` or `psa_ps_set_extended()` without invalidating the iteration context. 
```

etc... (there are other instances)

>    A caller may create a `uid` with ``psa_ps_set()`` or ``psa_ps_create()`` without invalidating the iteration context. However, the iterator is NOT guaranteed to return the new object, `uid`, the behaviour is dependent on both implementation and identity. In particular, the iterator is not expected to return `uid` if the iteration is already past the point at which it would naturally be returned. 

Reviewer: I think 'For example' works better than 'In particular' - as this is not normative text.

   
   A caller may call `psa_ps_rename(uid, uid_new)` without invalidating the iteration context. The iterator must not return `uid`. The iterator is not guaranteed to return `uid_new`, the behaviour is dependent on both implementation and identity.  

>    The following code snippet uses a linked list to store the matching files before iterating over that list and removing them. 

Reviewer: This comment is out of date as the linked list has been removed.

    * ``capacity`` unchanged.

    * Data in the ranges 0 to ``data_offset`` is not modified.

    * If ``data_offset + data_length < size`` then data in the range ``data_offset + data_length` to `size`` is not modified.

> .. function:: psa_ps_rename

Reviewer: Is there a reason why rename is not an API that we want to permit on the ITS storage?

   .. retval:: PSA_ERROR_NOT_SUPPORTED
        The implementation does not support the operation.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
        The operation failed because the physical storage has failed (Fatal error).

   The function renames ``uid`` to ``uid_new`` retaining the storage flags  that ``uid`` was created with.

   If the caller specifies `PSA_STORAGE_FLAG_REPLACE` the operation atomically replaces the existing contents of ```uid_new`` with those of ``uid``.

   Except in the case of ``PSA_ERROR_STORAGE_FAILURE``, in which case no guarantees can be made, the operation shall either succeed or leave storage unchanged.

.. function:: psa_ps_iterator_start

   .. summary::
       Initializes an iterator that can be used to return a list of uids in the Protected Storage.
       This function must be fully defined if iteration is supported.
       If iteration is not supported, then this function SHALL be defined in the header - potentially as an inline function - but SHALL return ``PSA_ERROR_NOT_SUPPORTED``

   .. param::  psa_ps_storage_iterator_t *context

>        The location in which the function should store a pointer to the newly created context for this iterator. 

Reviewer: ```suggestion
       The location in which the function stores the iteration context.
```

   A caller may delete a ``uid`` with `psa_ps_remove()` without invalidating the iteration context. the iterator MUST never return a ``uid`` that has been deleted. However, if the caller is multi-threaded it is possible another thread may delete a ``uid``.

   A caller may read the contents of any ``uid`` with `psa_ps_get()` or write with `psa_ps_set()` or `psa_ps_set_extended()` without invalidating the iteration context.

   A caller may create a ``uid`` with `psa_ps_set()` or `psa_ps_create()` without invalidating the iteration context. However, the iterator is NOT guaranteed to return the new object, ``uid``, the behavior is dependent on both implementation and identity. In particular, the iterator is not expected to return ``uid`` if the iteration is already past the point at which it would naturally be returned.

   A caller may call ``psa_ps_rename(uid, uid_new)`` without invalidating the iteration context. The iterator must not return ``uid``. The iterator is not guaranteed to return ``uid_new``, the behavior is dependent on both implementation and identity.

>    The following code snippet uses a linked list to store the matching files before iterating over that list and removing them.

Reviewer: Out-dated comment

   .. param:: psa_storage_uid_t *result
        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the next matching ``uid``. On error, the contents are undefined.

   .. return:: psa_status_t
       A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
       The operation completed successfully.

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
       The iterator has returned all the uids that match this iteration.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).

   .. retval:: PSA_ERROR_DATA_CORRUPT
       The operation failed because the contents of the iteration have changed. That is a ``uid`` matching the filter has either been created or deleted.

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
       The operation failed because either:

       * The provided context is not valid.

       * The caller cannot access the memory at ``result``

> 

Reviewer: Probably worth at least a single line in the function description to refer to the start function description of how the iterator works and how to use it.

.. function:: psa_ps_rename

   .. summary::
      Atomically renames the storage location with the specified ``uid`` to a ``uid_new``.

   .. param:: psa_storage_uid_t uid
        The current identifier for the data.

   .. param:: psa_storage_uid_t uid_new
        The new identifier for the data.

   .. param:: psa_storage_rename_flags_t rename_flags
        The flags must be either ``PSA_STORAGE_FLAG_NONE`` or ``PSA_STORAGE_FLAG_REPLACE``

   .. return:: psa_status_t
        A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
        The operation completed successfully.

   .. retval:: PSA_ERROR_ALREADY_EXISTS
        Storage with the specified ``uid_new`` already exists and ``rename_flags`` is `PSA_STORAGE_FLAG_NONE`

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
        Storage with the specified ``uid`` does not exist.

   .. retval:: PSA_ERROR_GENERIC_ERROR
        The operation failed because of an unspecified internal failure.

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
        The operation failed because either:

        * ``uid`` is ``0``.
        * ``uid_new`` is ``0``

>         * the ``psa_storage_rename_flags_t`` has a value set other than `PSA_STORAGE_FLAG_REPLACE`

Reviewer: ```suggestion
        * ``rename_flags`` has a value set other than `PSA_STORAGE_FLAG_REPLACE`.
```

   .. param::  psa_ps_storage_iterator_t* context
       A pointer to a context for this iterator as returned by `psa_ps_iterator_start` or updated by a previous call to `psa_ps_iterator_next`.  The content of the iterator will change on success and is undefined on error.

   .. param:: psa_storage_uid_t *result
        A pointer to the location in which to store ``uid``. On success the contents of this location will be updated with the next matching ``uid``. On error, the contents are undefined.

   .. return:: psa_status_t
       A status indicating the success or failure of the operation.

   .. retval:: PSA_SUCCESS
       The operation completed successfully.

   .. retval:: PSA_ERROR_DOES_NOT_EXIST
       The iterator has returned all the uids that match this iteration.

   .. retval:: PSA_ERROR_STORAGE_FAILURE
       The operation failed because the physical storage has failed (Fatal error).

   .. retval:: PSA_ERROR_INVALID_ARGUMENT
       The operation failed because either:

       * The provided context is not valid.

       * The caller cannot access the memory at ``result``

   Introduced in Version 1.1. 
   

>    This function takes an iterator that was returned by the `psa_ps_iterator_start` function and searches the storage for the next ``uid`` that matches the filters defined in that function call. If a ``uid`` matching the filter exists, the function updates the ``result`` and ``context`` parameters and retursn PSA_SUCCESS. If not it reruns an error. 

Reviewer: Spelling: 'reruns' -> 'returns'
